Tapas tf (#13393)
huggingface/transformers: docs/source/model_doc/tapas.rst

intermediate pre-training <https://www.aclweb.org/anthology/2020.findings-emnlp.27/>`__ by Julian Martin Eisenschlos,
Syrine Krichene and Thomas Müller.

> This model was contributed by `nielsr <https://huggingface.co/nielsr>`__. This model TF version was contributed by
> `kamalkraj <https://huggingface.co/kamalkraj>`__. The original code can be found `here

Reviewer: ```suggestion
This model was contributed by `nielsr <https://huggingface.co/nielsr>`__. The JAX/FLAX version of this model was contributed by
`kamalkraj <https://huggingface.co/kamalkraj>`__. The original code can be found `here
```

Author: Done

        >>> config = TapasConfig('google-base-finetuned-wikisql-supervised')
        >>> model = TapasForQuestionAnswering.from_pretrained('google/tapas-base', config=config)

> And here is the equivalent code for TensorFlow (To run the TF model install the `tensorflow_probability dependency
> <https://github.com/tensorflow/probability`>_ for your environment):

Reviewer: ```suggestion
In TensorFlow, this can be done as follows (make sure to have installed the `tensorflow_probability dependency
<https://github.com/tensorflow/probability`>_ for your environment):
```

Author: Done

        >>> from transformers import TapasConfig, TapasForQuestionAnswering

        >>> # you can initialize the classification heads any way you want (see docs of TapasConfig)

>         >>> config = TapasConfig(num_aggregation_labels=3, average_logits_per_cell=True)

Reviewer: Is there a reason you omitted this from the code example?

Author: Yes. Invalid combination of config parameters. with `select_one_column=False` model forward throw error. 
![Screenshot 2021-09-16 at 6 51 27 PM](https://user-images.githubusercontent.com/17096858/133619987-b24d2a3f-b971-4d8a-a6c2-da09f2a164ea.png)

        {'input_ids': tensor([[ ... ]]), 'attention_mask': tensor([[...]]), 'token_type_ids': tensor([[[...]]]),
        'numeric_values': tensor([[ ... ]]), 'numeric_values_scale: tensor([[ ... ]]), labels: tensor([[ ... ]])}

> change `return_tensors='tf'` in tokenizer to use with TF models.

Reviewer: ```suggestion
Set `return_tensors='tf'` when calling the tokenizer to use the TF models.
```

Author: Done

        ...             # add the float_answer which is also required (weak supervision for aggregation case)
        ...             encoding["float_answer"] = tf.convert_to_tensor(item.float_answer,dtype=tf.float32)
        ...             yield encoding['input_ids'], encoding['attention_mask'], encoding['numeric_values'], \
        ...                   encoding['numeric_values_scale'], encoding['token_type_ids'], encoding['labels'], \
        ...                   encoding['float_answer']
        ...
        ...     def __len__(self):
        ...        return len(self.data)

        >>> data = pd.read_csv(tsv_path, sep='\t')
        >>> train_dataset = TableDataset(data, tokenizer)
        >>> output_signature = (
        ... tf.TensorSpec(shape=(512,), dtype=tf.int32),
        ... tf.TensorSpec(shape=(512,), dtype=tf.int32),
        ... tf.TensorSpec(shape=(512,), dtype=tf.float32),
        ... tf.TensorSpec(shape=(512,), dtype=tf.float32),
        ... tf.TensorSpec(shape=(512,7), dtype=tf.int32),
        ... tf.TensorSpec(shape=(512,), dtype=tf.int32),
        ... tf.TensorSpec(shape=(512,), dtype=tf.float32))

>         >>> train_dataloader = tf.data.Dataset.from_generator(train_dataset,output_signature=output_signature).batch(32)

Reviewer: ```suggestion
        >>> train_dataloader = tf.data.Dataset.from_generator(train_dataset, output_signature=output_signature).batch(32)
```

Note that here, we encode each table-question pair independently. This is fine as long as your dataset is **not
conversational**. In case your dataset involves conversational questions (such as in SQA), then you should first group
together the ``queries``, ``answer_coordinates`` and ``answer_text`` per table (in the order of their ``position``
index) and batch encode each table with its questions. This will make sure that the ``prev_labels`` token types (see
docs of :class:`~transformers.TapasTokenizer`) are set correctly. See `this notebook
<https://github.com/NielsRogge/Transformers-Tutorials/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__

> for more info. see `this TF notebook
> <https://github.com/kamalkraj/Tapas-Tutorial/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__
> for more info using TF model.

Reviewer: ```suggestion
for more info. See `this notebook
<https://github.com/kamalkraj/Tapas-Tutorial/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__
for more info regarding using the TensorFlow model.
```

Author: Done

        ...         loss.backward()
        ...         optimizer.step()

> You can then fine-tune :class:`~transformers.TFTapasForQuestionAnswering` using native TensorFlow as follows (shown
> here for the weak supervision for aggregation case):

Reviewer: ```suggestion
Equivalently, in Tensorflow, fine-tuning :class:`~transformers.TFTapasForQuestionAnswering` can be done as follows (shown here for the weak supervision for aggregation case):
```

Author: Done

        ...        cell_values.append(table.iat[coordinate])
        ...     answers.append(", ".join(cell_values))

        >>> display(table)
        >>> print("")
        >>> for query, answer, predicted_agg in zip(queries, answers, aggregation_predictions_string):
        ...   print(query)
        ...   if predicted_agg == "NONE":
        ...     print("Predicted answer: " + answer)
        ...   else:
        ...     print("Predicted answer: " + predicted_agg + " > " + answer)    
        What is the name of the first actor?
        Predicted answer: Brad Pitt
        How many movies has George Clooney played in?
        Predicted answer: COUNT > 69
        What is the total number of movies?
        Predicted answer: SUM > 87, 53, 69

In case of a conversational set-up, then each table-question pair must be provided **sequentially** to the model, such
that the ``prev_labels`` token types can be overwritten by the predicted ``labels`` of the previous table-question
pair. Again, more info can be found in `this notebook
<https://github.com/NielsRogge/Transformers-Tutorials/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__.

> more info about TF model can be found in `this TF notebook
> <https://github.com/kamalkraj/Tapas-Tutorial/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__.

Reviewer: ```suggestion
more info about TF model can be found in `this TF notebook
<https://github.com/kamalkraj/Tapas-Tutorial/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__.
```

> pair. Again, more info can be found in `this notebook
> <https://github.com/NielsRogge/Transformers-Tutorials/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__.
> more info about TF model can be found in `this TF notebook
> <https://github.com/kamalkraj/Tapas-Tutorial/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__.

Reviewer: ```suggestion
pair. Again, more info can be found in `this notebook
<https://github.com/NielsRogge/Transformers-Tutorials/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__ (for PyTorch) and `this notebook <https://github.com/kamalkraj/Tapas-Tutorial/blob/master/TAPAS/Fine_tuning_TapasForQuestionAnswering_on_SQA.ipynb>`__ (for TensorFlow).
```

Author: Done
